docs/notebooks/recovering_from_errors.pct.py

```diff
@@ -0,0 +1,150 @@
+
+# %%
+class FaultyBranin:
+    def __init__(self):
+        self._is_broken = False
+
+    def manual_fix(self):
+        print("Go fix the machine")
+        self._is_broken = False
+
+    def __call__(self, x):
+        if self._is_broken:
+            raise Exception("Observer is broken")
+
+        if random.random() < 0.3:
+            self._is_broken = True
+
+        return {OBJECTIVE: trieste.data.Dataset(x, branin(x))}
+
+
+observer = FaultyBranin()
+
+# %% [markdown]
+# ## Set up the problem
+# We'll use the same set up as before, except for the acquisition rule, where we'll use `TrustRegion`, which will better illustrate how to recover.
+
+# %%
+search_space = trieste.space.Box(tf.constant([0.0, 0.0]), tf.constant([1.0, 1.0]))
+initial_data = observer(search_space.sample(5))
+
+variance = tf.math.reduce_variance(initial_data[OBJECTIVE].observations)
+kernel = gpflow.kernels.Matern52(variance, [0.2, 0.2]) + gpflow.kernels.White(1e-12)
+gpr = gpflow.models.GPR(astuple(initial_data[OBJECTIVE]), kernel, noise_variance=1e-5)
+gpflow.set_trainable(gpr.likelihood, False)
+models = {OBJECTIVE: trieste.models.GaussianProcessRegression(gpr)}
+
+trust_region = TrustRegion()
+
+# %% [markdown]
+# ## Run the optimization loop
+#
+# The optimizer returns an `OptimizationResult`. We can use that object's `astuple` method to get a 2-tuple of the final result and the history of the successful optimization steps.
+
+# %%
+bo = trieste.bayesian_optimizer.BayesianOptimizer(observer, search_space)
+
+result, history = bo.optimize(15, initial_data, models, trust_region).astuple()
+
+# %% [markdown]
+# ## Handling success
+#
+# We don't know if the optimization completed successfully or not, yet, so we'll only try to access and plot the data if it was successful. We can find out if this was the case with the `is_ok` attribute on the `Result`. If it was successful, we know there is data in the `result`, which we can `unwrap` and view.
+
+# %%
+if result.is_ok:
+    data = result.unwrap().datasets[OBJECTIVE]
+
+    arg_min_idx = tf.squeeze(tf.argmin(data.observations, axis=0))
+
+    print(f"query point: {data.query_points[arg_min_idx, :]}")
+    print(f"observation: {data.observations[arg_min_idx, :]}")
+
+# %% [markdown]
+# ## Handling failure
+#
+# If on the other hand, the optimization didn't complete successfully, we can fix our observer, and try again. We can try again by using the data, model and acquisition state from the last successful step, which is the last element of the `history`.
+#
+# Note we can view any `result` by printing it. We'll do that here to see what exception was caught.
+
+# %%
+if result.is_err:
+    print("result: ", result)
+
+    observer.manual_fix()
+
+    result, new_history = bo.optimize(
+        15 - len(history),
+        history[-1].datasets,
+        history[-1].models,
+        trust_region,
+        history[-1].acquisition_state
+    ).astuple()
+
+    history.extend(new_history)
```

Comment:
there an obvious method here
```python
class BayesianOptimizer:
    ...
    def recover(self, total_steps: int, history: List[Record[S]], acquisition_rule[S, SP]) -> OptimizationResult[S]:
       result, new_history = self.optimize(
            total_steps - len(history),
            history[-1].datasets,
            history[-1].models,
            acquisition_rule,
            history[-1].datasets
       )
       return OptimizationResult(result, history + new_history)
```
but I'll leave that for another PR

Reply:
is there an issue to keep track of this?

Reply:
#124